### PR TITLE
Get partial object implemented

### PR DESCRIPTION
**Pull request by chanchett**

### CODE REVIEW

File: src/ds3_net.h [hunk 1]
```diff
@@ -38,6 +38,7 @@ typedef struct {
 
 char* escape_url(const char* url);
 char* escape_url_object_name(const char* url);
+char* escape_url_range_header(const char* url);
```
Comment (rpmoore) [inline src/ds3_net.h]: I would write some unit tests that just test this method.  And throw a bunch of weird characters at it.

File: src/ds3.c [hunk 2]
```diff
@@ -428,6 +432,19 @@ void ds3_request_set_metadata(ds3_request* _request, const char* name, const cha
     g_free(prefixed_name);
 }
 
+void ds3_request_set_byte_range(ds3_request* _request, int64_t rangeStart, int64_t rangeEnd) {
+    char* range_value;
+    range_value = g_strdup_printf("bytes=%ld-%ld", rangeStart, rangeEnd);
```
Comment (rpmoore) [inline src/ds3.c]: We will have to think about this a little more.  You could have multiple ranges specified with a single request.

Comment (chanchett) [inline src/ds3.c]: I also thought would be a problem, but in the API it specifies that only one range is accepted.

File: src/ds3_net.h (hunk 1 above)
Comment (chanchett) [inline src/ds3_net.h]: Alright, should I do the same thing for escape_url_object_name as well?

File: src/ds3.c (hunk 2 above)
Comment (rpmoore) [inline src/ds3.c]: The API can accept more than one range.... At least I believe so, I'll check our integration tests on the Java SDK to be sure.

File: src/ds3_net.h (hunk 1 above)
Comment (rpmoore) [inline src/ds3_net.h]: Yeah, that would be a good idea.

File: src/ds3.c (hunk 2 above)
Comment (chanchett) [inline src/ds3.c]: Just looked it up. Our API says it works like the Amazon one does, and the _Amazon_ API only accepts one range. So ours might accept more. How would multiple ranges work with request length and offset?

Comment (rpmoore) [inline src/ds3.c]: The HTTP Range spec allows you to specify multiple ranges by separating them with commas.  So If you had two ranges you could do `Range: bytes=4-10,16-30`

Comment (chanchett) [inline src/ds3.c]: Right, but the Amazon API says it doesn't support that. I have no idea if
we follow what the Amazon API does or what HTTP spec says.

If we do want to support multiple ranges, the header bit wouldn't be too
hard, we could just check if the header exists, and if it does we append to
it. What would we do about the length and offset on the request?

Comment (rpmoore) [inline src/ds3.c]: I just confirmed that we support the behavior that I'm describing.  Length and offset stay the same on puts.  Gets only requires the offset.

File: src/ds3.c [hunk 3]
```diff
@@ -413,6 +417,20 @@ void ds3_request_set_metadata(ds3_request* _request, const char* name, const cha
     g_free(prefixed_name);
 }
 
+void ds3_request_set_byte_range(ds3_request* _request, int64_t rangeStart, int64_t rangeEnd) {
```
Comment (rpmoore) [inline src/ds3.c]: Should we provide a way to reset the Range header?

File: test/bulk_get.cpp [hunk 4]
```diff
@@ -122,6 +122,145 @@ BOOST_AUTO_TEST_CASE( bulk_get ) {
...
+    bool checksum_passed = true;
+
+    uint32_t segment_size = 3200;
+    for (i = 0; i < chunk_response->object_list->list_size; i++) {
+        ds3_bulk_object_list* chunk_object_list = chunk_response->object_list->list[i];
+        for(n = 0; n < chunk_object_list->size; n++, file_index++) {
+            FILE* w_file;
+            ds3_bulk_object current_obj = chunk_object_list->list[n];
+            request = ds3_init_get_object_for_job(bucket_name, current_obj.name->value, current_obj.offset, bulk_response->job_id->value);
+	    ds3_request_set_byte_range(request, segment_size, segment_size*2-1);
```
Comment (rpmoore) [inline test/bulk_get.cpp]: the indentation looks off here

File: test/checksum.cpp [hunk 5]
```diff
@@ -66,3 +66,7 @@ bool compare_hash(char* filename_1, char* filename_2) {
     return passed;
 }
 
+// Function which compares checksums of the files passed
+bool compare_hash(char* filename_1, char* filename_2) {
+  return compare_hash_extended(filename_1, filename_2, 0, 0, 0);
```
Comment (rpmoore) [inline test/checksum.cpp]: this should be 4 spaces instead of 2

File: src/ds3.c (hunk 3 above)
Comment (chanchett) [inline src/ds3.c]: Yeah, probably. Added a reset_byte_range call.

**Pull request merged**